### PR TITLE
Update samples to MSAL v4

### PR DESCRIPTION
**Pull request by BrianTJackett**
Implements #48 
Updates daily apps to MSAL v4 and also update README files to reflect changed code.

Also includes start of sample for certificate authentication with base console app.

### CODE REVIEW

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    // This class encapsulates the details of getting a token from MSAL and exposes it via the 
+    // IAuthenticationProvider interface so that GraphServiceClient or AuthHandler can use it.
+    // A significantly enhanced version of this class will in the future be available from
+    // the GraphSDK team.  It will supports all the types of Client Application as defined by MSAL.
+    public class MsalAuthenticationProvider : IAuthenticationProvider
+    {
+        private IConfidentialClientApplication _clientApplication;
+        private string[] _scopes;
+
+        public MsalAuthenticationProvider(IConfidentialClientApplication clientApplication, string[] scopes) {
```
Comment (jmprieur) [inline base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: This supposes that you don't change the scopes later?
this is ok for a sample, but does it scale to real life apps?

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            request.Headers.Authorization = new AuthenticationHeaderValue("bearer", token);
+        }
+
+        /// <summary>
+        /// Acquire Token 
+        /// </summary>
+        public async Task<string> GetTokenAsync()
+        {
+            AuthenticationResult authResult = null;
+            authResult = await _clientApplication.AcquireTokenForClient(_scopes)
```
Comment (jmprieur) [inline base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: Shouln't the the sample be named dotnetcore-console-daemon-sample ? as this is a daemon app only?

File: base-console-app-cert/Program.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            List<QueryOption> options = new List<QueryOption>
+            {
+                new QueryOption("$top", "1")
+            };
+
+            var graphResult = graphClient.Users.Request(options).GetAsync().Result;
+            Console.WriteLine("Graph SDK Result");
+            Console.WriteLine(graphResult[0].DisplayName);
+
+            //Direct query using HTTPClient (for beta endpoint calls or not available in Graph SDK)
```
Comment (jmprieur) [inline base-console-app-cert/Program.cs]: Why show both approaches? is it to show that customers can use the Graph SDK or REST APIs?

File: base-console-app-cert/Program.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            var authority = $"https://login.microsoftonline.com/{config["tenantId"]}/v2.0";
+
+            // defaulting to CurrentUser certificate store under My (Personal), change these if stored elsewhere
+            X509Certificate2 cert = GetCertificate(certificateThumbprint, StoreName.My, StoreLocation.CurrentUser);
+
+            //this specific scope means that application will default to what is defined in the application registration rather than using dynamic scopes
+            List<string> scopes = new List<string>();
+            scopes.Add("https://graph.microsoft.com/.default");
+
+            var cca = ConfidentialClientApplicationBuilder.Create(clientId)
```
Comment (jmprieur) [inline base-console-app-cert/Program.cs]: Why not have used `CreateWithOptions()` ? (with a slightly different format of the appsettings.json)

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs (hunk 1 above)
Comment (BrianTJackett) [reply base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: @jmprieur By changing scopes do you mean 1) dynamic consent to additional scopes or 2) changing administratively assigned scopes on the app registration?  For the former, the majority of samples in this repo use app-only permissions in a non-interactive session thus dynamic consent doesn't apply to those scenarios.  For the latter the README files for each app detail the permissions required for each sample and are not expected to change.  If I misunderstood please let me know.

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs (hunk 2 above)
Comment (BrianTJackett) [reply base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: Yes, the majority of samples in this repo are daemon apps.  Originally one of our blog series contributing authors was going to include an additional 3-4 interactive samples but they ended up being hosted on a different repo.  We could rename this repo but I'll have to look into the impact of refactoring any links, blog posts, and other internal references.

File: base-console-app-cert/Program.cs (hunk 3 above)
Comment (BrianTJackett) [reply base-console-app-cert/Program.cs]: Some of the later daily samples call endpoints that are not available through the Graph SDK (at least at the time of writing).  Additionally we wanted to show readers both options for making authenticated Graph calls.

File: base-console-app-cert/Program.cs (hunk 4 above)
Comment (BrianTJackett) [reply base-console-app-cert/Program.cs]: @jmprieur Great suggestion.  The least amount of changes were implemented to update to the MSAL .Net v3 / v4 APIs.  Refactoring the original appsettings.json will take additional effort across each of the impacted projects.  Please open an issue (or I can if you'd like) to implement those changes.

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs (hunk 1 above)
Comment (jmprieur) [reply base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: Good point that this is a daemon. In that case you could just use `https://graph.microsoft.com/.default`

File: base-console-app-cert/Helpers/MsalAuthenticationProvider.cs (hunk 2 above)
Comment (jmprieur) [reply base-console-app-cert/Helpers/MsalAuthenticationProvider.cs]: GitHub does a good job of maintaining the URL to the previous names of the repos. So links would not be a problem. But I see your point.

File: base-console-app-cert/Program.cs (hunk 3 above)
Comment (jmprieur) [reply base-console-app-cert/Program.cs]: thanks for the update. This makes sense.

File: base-console-app-cert/Program.cs (hunk 4 above)
Comment (jmprieur) [reply base-console-app-cert/Program.cs]: https://github.com/microsoftgraph/dotnetcore-console-sample/issues/51

**Pull request merged**